rfcs/006-package-management.md

```diff
@@ -0,0 +1,359 @@
+interface is more similar to `cargo`).
+
+## Kinds of dependencies
+
+Where can dependencies come from? Dhall allows imports from arbitrary urls. Nix
+supports fetching from a variety of VCSs, paths, and archive formats.
+
+We'll support dependencies from
+
+- a central registry, that can identify packages by name and version number.
+  This should be the most common method of importing packages, like `crates.io`
+  in rust.
+- git repositories (either from HEAD, or from branches, tags, and revisions
+  specified by hashes). This allows for easy use of unpublished packages,
+  including in-development versions.
+- paths (relative or absolute). This allows for easy use of different packages
+  within the same repository, or for temporary patching of published packages.
+
+We require packages to have their own manifest file (at the package root), even
+if they don't import dependencies.
+
+## Lock files
+
+In order to ensure reproducibility across time and across machines, we build
+a lock-file (if there isn't yet one) when running `nickel eval` or `nickel
+export`. The lock-file specifies the exact versions of all (transitive)
+dependencies, allowing those identical versions to be used every time.
+
+- For a git dependency, the manifest might not specify the exact revision (it
+  might specify a branch or tag, or just default to HEAD). The lock-file will
+  record the exact revision.
+- For a repository dependency, the version specifier might allow for a range of
+  versions. The lock-file will record the exact version used.
+- For a path dependency, the lock-file will record that there was a path
+  dependency, but it won't record anything about it and it will ignore recursive
+  dependencies. This is because path dependencies can change at any time, so
+  they can't be meaningfully locked.
+
+## Restrictions on path dependencies
+
+Path dependencies can be problematic for reproducibility, because they require
+something to be present at a given path. In order to mitigate this:
+
+- Packages imported from the index are not allowed to have path dependencies.
+- Packages imported from git can have path dependencies, but only if they point
+  within the same git repo. The lock-file treats these path dependencies as though
+  they were git dependencies.
+
+Therefore your dependency tree can have a few path-dependency subtrees, but only
+at the root.
+
+## Stale lock files
+
+What happens if there's a lock-file, but someone modifies the manifest? The
+lock-file might need to be regenerated: certainly it might need some new
+entries, but also there might be new version conflicts that require a different
+resolution. In this case, we treat the lock-file as a suggestion instead of a
+hard constraint: during resolving, when choosing the next package version to
+try, we try to pick the locked version first. But if we run into a resolution
+conflict, we allow a different version to be chosen (and notify the user
+that a package was changed).
```

Comment:
An alternative is to never do any dependency resolution at all: target a particular commit of the registry, and just use the latest version of each package according to said commit. This is the Nixpkgs way (and also Stackage). It has pros and cons, but it has a fairly clear story for lockfile updates (_i.e._ just create lockfile entries for the new packages, never touch the existing package (unless asked to by the user, of course)).

Reply:
The lack of dependency resolution in flakes leads people to implement their own layer (https://discourse.nixos.org/t/introducing-flakehub/32044) which the usual lot of community split, bikeshedding and drama. I don't know about the Stack situation, though.

Reply:
I think that the flakes and Stack models rely on the existence of a globally coherent versioned set of packages (is that what Stackage is)? In Nixpkgs those are NixOS releases. I agree this is usually great from a user experience point of view, but also a non negligible burden put on the people in charge of curating, updating and releasing this set.

Reply:
Yes, Stackage is a collection of so-called snapshots. Each snapshot has at most one version of each package, which are tested to be able to build together.

I think one of the biggest obstacle that you'd have for Nickel is probably the testing bit. Since you don't necessarily have types to guarantee that everything works together well (and evaluating everything like Nixpkgs does is pretty prohibitive).

Possibly the approach of allowing several copies of the same package alleviates a large chunk of the issues with bounds and resolution. But I think this alternative should at least be addressed in the document.

PS: if we go with version bounds, we need to have an answer to the question: what is the process when we discover after the fact that a bound is too permissive?

Reply:
I added an "Alternative" section about global snapshots as a placeholder, but it's probably worth discussing it in more detail in the weekly.

Anecdotally, I find that the rust defaults (dependencies use the "^" mode by default, and you can have multiple copies of a package that don't share the same semver bin) work pretty well. The point is that the default dependency bounds align with the no-duplication bounds, and so if those bounds are too restrictive then you just end up with multiple copies of a package. 

The bounds are usually not too permissive, because they are then someone has done semver wrong. But if it does happen, you can often work around it by pinning a dependency. For example, suppose I have a dependency on `dep`, which in turn depends on `dep2 ^0.1.0`, but then `dep2` releases a broken `0.1.1`. I can force version resolution to fall back to the old version of `dep2` by adding a dependency on `dep2 =0.1.0`.

Reply:
I'm sorry, my question was ambiguous. Here's the scenario that I have in mind: we have packages A<-B uploaded in the registry. And we discover that the latest version of B, says it's compatible with versions of A which it actually isn't compatible with (whatever the reason). We now need to update the registry somehow. It could be by altering the how the current version of B depends on A. Or it can be by marking this version of B as incorrect somehow, so that it's never picked up by the solver (and uploading a new version).

But, one way or another, we need to make sure that the bad combination is forbidden, otherwise, it'll be possible in the future for the solver to pick that combination, even if there are newer versions of B.

We need policy and workflows for to handle this case.

Reply:
Ah, right. I'll add a section on that. I like the "mark B as incorrect" option, because it can also be used to work around other kinds of breakage.